README.md
# rb175_file_based_cms
This is a project for the Launch School course RB 175.


data/changes.txt
The Ruby Changes

This site is dedicated to history of Ruby language evolution. Basically, it is just the same information that each Ruby version’s NEWS file contains, just in more readable and informative manner.

    Latest version: 3.0 (2.7 − 2.6 −2.5 − 2.4)

Main goals:

    Full: unlike most of “What’s new in Ruby x.y?” blog posts, information here targets to cover all the NEWS file of current Ruby version;
    Comprehensive: unlike the NEWS file itself (and most of blog-posts, too), the site provides full context for each change:
        Where and how it was discussed;
        Related documentation at ruby-doc.org;
        Code examples;
        Reasoning for the change, if known.
    Concise: given two goals above, content still tries to stay short and focused, so the changes could be quickly navigated through;
    Well-structured: both regarding order/explanations of each particular change, and design of the site, it is intended to be easily and logically navigated.
    Open: the source of changelog is available on the GitHub and is open for fixes and suggestions.

Some things to know about the content:

    Structure and grouping of the information, as well as the illustrative examples, are mostly decided by this site’s author. I am trying to make it as useful and logical as humanly possible, but I have my own opinions about what is important, and what is not, what is related and what is not.
        Maybe “Highlights” section at the top of each version changelog is more subjective than the rest of the content, but it is totally optional, you can skip it.
    The site is dedicated to the language, not its implementation(s), therefore at the moment it does not include the description of MRI implementation changes, optimizations and internals.
    …and also some minor behavior changes are excluded.
    I want to eventually cover Ruby versions down to 1.8, or maybe even earlier, but it is currently work-in-progress, with the first priority of the recent release of Ruby 2.6, and then going down version by version in my free time.
    UPD 2019-06-06: Ruby 2.5 is now covered too. Despite being 1.5 years old news, I believe it is still important to cover it in the same manner as the recent version was.
    UPD 2019-10-14: Ruby 2.4 changelog added, and some other content changed. See History for detail.
    UPD 2019-12-27: Newly released Ruby 2.7 changelog added.
    UPD 2019-12-25: Newly released Ruby 3.0 changelog added.

The source of the site can be found at GitHub. See also the Contributing section.

test/cms_test.rb
ENV["RACK_ENV"] = "test"

require "minitest/autorun"
require "rack/test"
require "fileutils"

require_relative "../cms.rb"

def sign_in(user='admin', pass='secret')
  post "/users/signin", username: user, password: pass
end

def sign_out
  get '/users/signout'
end

class AppTest < Minitest::Test
  include Rack::Test::Methods

  def app
    Sinatra::Application
  end

  def setup
    sign_in
    FileUtils.mkdir_p(root)
    create_document "about.md", "#Ruby is simple in appearance, but is very complex inside"
    create_document "another_document.txt"
    create_document "changes.txt", "This site is dedicated to history of Ruby language evolution. Basically, it is just the same information that each Ruby versionâ€™s NEWS file contains, just in more readable and informative manner."
  end

  def teardown
    sign_out
    FileUtils.rm_rf(root)
  end

  def test_sign_in_page
    get '/users/signin'

    assert_equal 200, last_response.status
    assert_includes last_response.body, 'action="/users/signin" method="post"'
    assert_includes last_response.body, 'label for="username">Username'
    assert_includes last_response.body, 'label for="password">Password'
    assert_includes last_response.body, 'type="submit" value="Sign In"'
  end

  def test_sign_in
    sign_out
    sign_in('admin', 'secret')

    assert_equal 302, last_response.status

    get last_response['Location']

    assert_equal 200, last_response.status
    assert_includes last_response.body, 'Welcome!'
    assert_includes last_response.body, 'about.md'
  end

  def test_sign_in_invalid_credentials
    sign_out
    sign_in('invalid', 'password')

    assert_equal 422, last_response.status
    assert_includes last_response.body, 'Invalid credentials.'
    assert_includes last_response.body, 'action="/users/signin" method="post"'
    assert_includes last_response.body, 'label for="username">Username'
    assert_includes last_response.body, 'label for="password">Password'
    assert_includes last_response.body, 'type="submit" value="Sign In"'
  end

  def test_sign_out
    sign_out

    assert_equal 302, last_response.status

    get last_response["Location"]

    assert_includes last_response.body, "You have been signed out."
    assert_includes last_response.body, 'action="/users/signin"'
  end

  def test_index_signed_out
    sign_out
    get "/"

    assert_equal 200, last_response.status
    assert_includes last_response.body, 'action="/users/signin"'
    refute_includes last_response.body, 'about.md'
  end

  def test_index_signed_in # test for listing of documents, edit links, new document
    get "/"

    assert_equal 200, last_response.status
    assert_includes last_response["Content-Type"], "text/html"
    assert_includes last_response.body, "about.md"
    assert_includes last_response.body, "changes.txt"
    assert_includes last_response.body, 'action="about.md/edit'
    assert_includes last_response.body, '<a href="/new_doc/">'
    assert_includes last_response.body, 'action="about.md/delete"'
    assert_includes last_response.body, 'Signed in as admin'
    assert_includes last_response.body, 'action="/users/signout"'
    assert_includes last_response.body, 'Sign Out</button>'
  end
 
  def test_content # test to see if content can be accessed
    get "/changes.txt"

    assert_equal 200, last_response.status
    assert_includes last_response["Content-Type"], "text/plain"
    assert_includes last_response.body, "This site is dedicated to history of Ruby language"
  end

  def test_nonexistant_content # test for redirection, error message, clearing of error message
    get "/thisfiledoesnotexist.txt"

    assert_equal 302, last_response.status

    get last_response["Location"]

    assert_equal 200, last_response.status
    assert_includes last_response.body, "thisfiledoesnotexist.txt does not exist."

    get "/"
    refute_includes last_response.body, "thisfiledoesnotexist.txt does not exist."
  end

  def test_markdown_to_html # test for content conversion from markdown to HTML
    get "/about.md"

    assert_equal 200, last_response.status
    assert_includes last_response["Content-Type"], "text/html"
    assert_includes last_response.body, "<h1>Ruby is"
  end

  def test_content_edit_page # test contents of edit page
    get "/changes.txt/edit"

    assert_equal 200, last_response.status
    assert_includes last_response.body, "Edit content of changes.txt:"
    assert_includes last_response.body, "<textarea"
    assert_includes last_response.body, 'type="submit"'
  end

  def test_nonexistant_content_edit_page
    get "/thisfiledoesnotexist.txt/edit"

    assert_equal 302, last_response.status
    
    get last_response["Location"]

    assert_equal 200, last_response.status
    assert_includes last_response.body, "thisfiledoesnotexist.txt does not exist."

    get "/"
    refute_includes last_response.body, "thisfiledoesnotexist.txt does not exist."
  end

  def test_content_update
    new_content = "THIS IS A NEW EDIT VIA TESTING - #{Time.now.getutc}"

    post "/changes.txt/edit", updated_content: new_content

    assert_equal 302, last_response.status

    get last_response["Location"]
    
    assert_equal 200, last_response.status
    assert_includes last_response.body, "changes.txt has been updated."

    get "/changes.txt"
    assert_equal 200, last_response.status
    assert_includes last_response.body, new_content
  end

  def test_new_content_page
    get "/new_doc/"

    assert_equal 200, last_response.status
    assert_includes last_response.body, "Add a new document:"
    assert_includes last_response.body, '<input type="text"'
    assert_includes last_response.body, '<input type="submit"'
  end

  def test_content_creation
    new_doc = "new.txt"
    sign_in

    post "/new_doc/", doc_name: new_doc
    
    assert_equal 302, last_response.status
    
    get last_response["Location"]

    assert_equal 200, last_response.status
    assert_includes last_response.body, "#{new_doc} was created."
    assert_includes last_response.body, "href=\"#{new_doc}\""
  end

  def test_content_creation_no_name
    post "/new_doc/", doc_name: ""

    assert_equal 422, last_response.status
    assert_includes last_response.body, "A name is required."
  end

  def test_content_creation_invalid_type
    post "/new_doc/", doc_name: "invalid.invalid"
    
    assert_equal 415, last_response.status
    assert_includes last_response.body, "The file must be #{SUPPORTED_TYPES.join(' or ')} file types."
  end

  def test_content_creation_no_type
    post "/new_doc/", doc_name: "invalid"
    
    assert_equal 415, last_response.status
    assert_includes last_response.body, "The file must be #{SUPPORTED_TYPES.join(' or ')} file types."
  end

  def test_content_creation_already_exists
    post "/new_doc/", doc_name: "about.md"
    
    assert_equal 409, last_response.status
    assert_includes last_response.body, "about.md already exists."
  end

  def test_content_deletion
    post "/about.md/delete"

    assert_equal 302, last_response.status

    get last_response["Location"]
    
    assert_equal 200, last_response.status
    assert_includes last_response.body, "about.md was deleted."
    refute_includes last_response.body, '<a href="about.md">'
  end
end


cms.rb
# frozen_string_literal: true

require 'sinatra'
require 'sinatra/reloader' if development?
require 'sinatra/content_for'
require 'tilt/erubis'
require 'redcarpet'

SUPPORTED_TYPES = ['.txt', '.md']
VALID_CREDENTIALS = [ 
  ['admin', 'secret'] 
]

def render_markdown(string)
  markdown = Redcarpet::Markdown.new(Redcarpet::Render::HTML)
  markdown.render(string)
end

def get_content(path)
  File.read(path)
end

def write_content(path, new_content)
  File.write(path, new_content)
end

def load_document(path)
  content = get_content(path)

  case File.extname(path).downcase
  when ".txt"
    headers["Content-Type"] = "text/plain"
    content
  when ".md"
    erb render_markdown(content), layout: :layout
  end
end

def validate_document_access(path)
  if File.file?(path)
    yield
  else
    session[:message] = "#{params[:filename]} does not exist."
    redirect "/"
  end
end

def file_path(filename)
  File.join(root, params[:filename])
end

def root
  if ENV["RACK_ENV"] == "test"
    File.expand_path("../test/data", __FILE__)
  else
    File.expand_path("../data", __FILE__)
  end
end

def create_document(name, content = "")
  File.open(File.join(root, name), "w") do |file|
    file.write(content)
  end
end

def validate_document_creation(path)
  new_doc = File.basename(path)

  if File.file?(path)
    status 409
    session[:message] = "#{new_doc} already exists."
    erb :new_doc, layout: :layout
  else
    create_document(new_doc)
    session[:message] = "#{new_doc} was created."
    redirect "/"
  end
end

def supported_doc_type?(filename)
  SUPPORTED_TYPES.include?(File.extname(filename).downcase)
end

def valid_credentials?(username, password)
  VALID_CREDENTIALS.include?([username, password])
end

def validate_user
  valid_users = VALID_CREDENTIALS.map { |cred| cred[0] }
  if valid_users.include?(session[:username])
    yield
  else
    invalid_user_force_sign_in
  end
end

def invalid_user_force_sign_in
  session.delete(:username)
  status 422
  session[:message] = 'Invalid credentials.'
  erb :sign_in, layout: :layout
end

helpers do
  def display_message
    session.delete(:message) if session[:message]
  end
end

configure do
  set :erb, :escape_html => true
  enable :sessions
  set :session_secret, 'secret'
end

before do
  @docs = Dir.glob(root + '/*').map do |path|
    File.basename(path) if File.ftype(path) == 'file'
  end.compact
end

get '/' do
  if session[:username].nil?
    erb :index, layout: :layout
  else
    erb :doc_list, layout: :layout
  end
end

get '/users/signin' do
  erb :sign_in, layout: :layout
end

post '/users/signin' do
  if valid_credentials?(params[:username], params[:password])
    session[:username] = params[:username]
    session[:message] = 'Welcome!'
    redirect '/'
  else
    invalid_user_force_sign_in
  end
end

get '/users/signout' do
  session.delete(:username)
  session[:message] = 'You have been signed out.'
  redirect '/'
end

get "/new_doc/" do
  erb :new_doc, layout: :layout
end

post "/new_doc/" do
  validate_user do
    new_doc = params[:doc_name]

    if new_doc == ""
      status 422
      session[:message] = "A name is required."
      erb :new_doc, layout: :layout
    elsif !supported_doc_type?(new_doc)
      status 415
      session[:message] = "The file must be #{SUPPORTED_TYPES.join(' or ')} file types."
      erb :new_doc, layout: :layout
    else new_doc == ""
      validate_document_creation(File.join(root, new_doc))
    end
  end
end

get "/:filename" do # look at a document
  validate_user do
    path = file_path(params[:filename])

    validate_document_access(path) { load_document(path) }
  end
end

get "/:filename/edit" do # edit a document
  validate_user do
    path = file_path(params[:filename])

    validate_document_access(path) do
      @content = get_content(path)
      erb :doc_edit, layout: :layout
    end
  end
end

post "/:filename/edit" do # submit edits to a document
  validate_user do
    write_content(file_path(params[:filename]), params[:updated_content])

    session[:message] = "#{params[:filename]} has been updated."
    redirect "/"
  end
end

post "/:filename/delete" do
  validate_user do
    path = file_path(params[:filename])

    validate_document_access(path) do
      File.delete(path)
      session[:message] = "#{params[:filename]} was deleted."
      redirect "/"
    end
  end
end